Datasets for benchmarking strategies

paulmorio opened the pull request:
Modules to download and process datasets from online sources into`torch.utils.data.Dataset` instances, with additional attributes for (stratified) k-fold CV as described in the paper.

This incurs a few new dependencies namely `openpyxl`, `xlrd`, and `pyreadr` for processing the excel and R data storage formats for original raw datas. 

Also included are utility functions for transforming each of the datasets into datamanagers that have "cold" or "warm" label initialisations for benchmarking AL strategies on the datasets.

Includes tests for all the modules implemented

thomasgaudelet reviewed (commented):
LGTM! Small things on the side
1) could you update the coverage on the readme? (it's done manually at the moment as I haven't had time to crack how to have it done automatically....)
2) could you add a page in documentation for the datasets?

paulmorio commented:
Updated with some tests for `uci_datasets` I looked into updating the coverage dynamically and the best solution I've come across so far is described here for github actions: https://github.com/marketplace/actions/dynamic-badges

Unfortunately I don't have the rights to access the secrets settings for the repo so I can't finish the instructions there.

I've added a reference documentation page. This could be followed by another tutorial at a later date.

thomasgaudelet reviewed (approved):
Nice!

jyperion commented on .pre-commit-config.yaml:
```diff
@@ -6,7 +6,7 @@ repos:
     -   id: trailing-whitespace
 -   repo: https://github.com/psf/black
-    rev: 21.12b0
+    rev: 22.3.0
```
Is there a need for this bump?

jyperion commented on pyrelational/datasets/uci_datasets.py:
```diff
@@ -0,0 +1,194 @@
+        elif self.name == "power":
+            zipfile.ZipFile(self.data_dir + "UCI/CCPP.zip").extractall(self.data_dir + "UCI/")
+            data = pd.read_excel(self.data_dir + "UCI/CCPP/Folds5x2_pp.xlsx", header=0).values
+            np.random.shuffle(data)
```
Why the shuffle here?

paulmorio replied:
You are right this is redundant. Fixed in latest commit.

jyperion commented on pyrelational/datasets/uci_datasets.py:
```diff
@@ -0,0 +1,194 @@
+            urllib.request.urlretrieve(self.datasets[self.name], self.data_dir + "UCI/" + file_name)
+        data = None
+
+        if self.name == "housing":
```
Can there be a cleaner way than having these many if statements?

paulmorio replied:
Some of the datasets require some additional processing after downloading, so I thought this was the most straightforward way, I am not aware of a better way to handle this, cases would look the same, delegating it to another module to handle the resolution of the conditionals also would just be more code I think. I can't immediately think of a better way atm.
